### ISSUE TITLE
NiftiLabelsMasker returns biased time series compared to NiftiMasker when Standardize=True

### ISSUE THREAD

**Issue by surchs** [nilearn/nilearn#1954]
Hi,

__What version of Nilearn are you using__
```
nilearn.__version__
'0.5.0'
```

__What were you trying to do?__
Extract standardized voxel and region timeseries using `NiftiMasker` and `NiftiLabelsMasker` respectively.

__What did you expect will happen?__
I expect the returned voxel timeseries to be identical to a corresponding one-voxel region time series.

__What actually happened?__
There is a small but consistent difference between the voxel and region timeseries if `standardize=True` but not if `standardize=False`.

__This code reproduces my observation__
```
import numpy as np
import nibabel as nib
from nilearn import datasets, input_data

# Get some data
data = datasets.fetch_adhd(n_subjects=1)
fmri_p = data.func[0]
func_i = nib.load(fmri_p)

# Make a mask from the functional data
mask = np.std(func_i.get_data(), 3) > 1
mask_i = nib.Nifti1Image(mask.astype(int), affine=func_i.affine, header=func_i.header)

# Create a dummy atlas with a single voxel ROI
dummy_atlas = np.zeros(mask.shape)
dummy_vec = np.zeros(np.sum(mask))
dummy_vec[0] = 1
dummy_atlas[mask] = dummy_vec
dummy_atlas_i = nib.Nifti1Image(dummy_atlas, affine=mask_i.affine, header=mask_i.header)

# Define standardized and non-standardized maskers
label_masker_stand = input_data.NiftiLabelsMasker(labels_img=dummy_atlas_i, mask_img=mask_i, standardize=True)
voxel_masker_stand = input_data.NiftiMasker(mask_i, standardize=True)

label_masker = input_data.NiftiLabelsMasker(labels_img=dummy_atlas_i, mask_img=mask_i, standardize=False)
voxel_masker = input_data.NiftiMasker(mask_i, standardize=False)

# Extract the corresponding voxel and ROI timeseries
voxel_ts_stand = voxel_masker_stand.fit_transform(func_i)
dummy_ts_stand = label_masker_stand.fit_transform(func_i)

voxel_ts = voxel_masker.fit_transform(func_i)
dummy_ts = label_masker.fit_transform(func_i)

# Report the differences
print('raw value SSE: ', np.sum(np.square(voxel_ts[:, 0] - dummy_ts.squeeze())))
print('standardized value SSE: ', np.sum(np.square(voxel_ts_stand[:, 0] - dummy_ts_stand.squeeze())))

# Show the differences
from matplotlib import pyplot as plt
plt.plot(voxel_ts[:, 0] - dummy_ts.squeeze(), label='raw values')
plt.plot(voxel_ts_stand[:, 0] - dummy_ts_stand.squeeze(), label='standardized values');
plt.legend();
```

__This is what this code generates__
raw value SSE:  0.0
standardized value SSE:  1.2196836e-09
![image](https://user-images.githubusercontent.com/1302022/55279317-f2752100-52ed-11e9-9954-0ec02c90d413.png)

**Comment by KamalakerDadi** [comment]
On a first look, I tried to plot the dummy_atlas_i on brain. It doesn't seem like point is overlaying with the brain. May be I understood the issue wrong ?
![dummy_atlas_i](https://user-images.githubusercontent.com/11410385/55280134-9af0a880-5321-11e9-8935-f0e0d3caabb0.png)

**Comment by surchs** [comment]
Hey @KamalakerDadi, yeah I in fact didn't check for that, sorry about that. However it doesn't change the oberservation. If you run this instead, you still get a difference:
```
mask = np.std(func_i.get_data(), 3) > 1
mask_i = nib.Nifti1Image(mask.astype(int), affine=func_i.affine, header=func_i.header)
dummy_atlas = np.zeros(mask.shape)
dummy_atlas[tuple(map(lambda x: int(np.floor(x/2)), dummy_atlas.shape))] = 1
dummy_atlas_i = nib.Nifti1Image(dummy_atlas, affine=mask_i.affine, header=mask_i.header)

label_masker_stand = input_data.NiftiLabelsMasker(labels_img=dummy_atlas_i, mask_img=mask_i, standardize=True)
voxel_masker_stand = input_data.NiftiMasker(mask_i, standardize=True)

label_masker = input_data.NiftiLabelsMasker(labels_img=dummy_atlas_i, mask_img=mask_i, standardize=False)
voxel_masker = input_data.NiftiMasker(mask_i, standardize=False)

voxel_ts_stand = voxel_masker_stand.fit_transform(func_i)
dummy_ts_stand = label_masker_stand.fit_transform(func_i)

voxel_ts = voxel_masker.fit_transform(func_i)
dummy_ts = label_masker.fit_transform(func_i)

vox_ind = np.where(dummy_atlas[mask])[0][0]
plt.plot(voxel_ts[:, vox_ind] - dummy_ts.squeeze(), label='raw values')
plt.plot(voxel_ts_stand[:, vox_ind] - dummy_ts_stand.squeeze(), label='standardized values');
plt.legend();
print('raw value SSE: ', np.sum(np.square(voxel_ts[:, vox_ind] - dummy_ts.squeeze())))
print('standardized value SSE: ', np.sum(np.square(voxel_ts_stand[:, vox_ind] - dummy_ts_stand.squeeze())))
```
raw value SSE:  0.0
standardized value SSE:  2.2377595e-12
![image](https://user-images.githubusercontent.com/1302022/55280348-fc515100-52fa-11e9-9511-cdbe88dc56fd.png)

but now the roi is bang in the center:
![image](https://user-images.githubusercontent.com/1302022/55280359-10954e00-52fb-11e9-82da-0f2a3f74d5c9.png)

**Comment by KamalakerDadi** [comment]
My reasoning may not makes sense. But, I would think that the difference is coming from standardization i.e. variance normalization based on single voxel and multiple voxels ?
May be @bthirion has some intuitions.

**Comment by surchs** [comment]
But the ROI is only one voxel as well. Maybe the standardization takes into account signals beyond the ROI or potentially in the mask?

**Comment by surchs** [comment]
I think this might be just a precision issue:
```
print(np.std(dummy_ts_stand), np.mean(dummy_ts_stand), dummy_ts_stand.dtype)
1.0 -2.080744e-06 float32
```
The deviations are consistent with that level of precision. And I guess that there is a good reason for why these operations are happening with 32 bit floats. So I think we can close this.

**Comment by bthirion** [comment]
Yes, I think it's nothing else than numerical precision.
